ssl/statem/extensions.c

```diff
@@ -857,14 +857,6 @@ static int init_status_request(SSL *s, unsigned int context)
 {
     if (s->server) {
         s->ext.status_type = TLSEXT_STATUSTYPE_nothing;
-    } else {
-        /*
-         * Ensure we get sensible values passed to tlsext_status_cb in the event
-         * that we don't receive a status message
-         */
-        OPENSSL_free(s->ext.ocsp.resp);
-        s->ext.ocsp.resp = NULL;
-        s->ext.ocsp.resp_len = 0;
```

Comment:
We probably need to modify the tls_collect_extensions() and tls_parse_all_extensions() functions in ssl/statem/extensions.c.

tls_collect_extensions(), amongst other things, is responsible for calling the init functions for all the extensions. tls_parse_all_extensions() calls all the parse functions and the finalisers. I think we need to modify the current behaviour so that init and final are called once *per message*, while parse is called once per extension block (assuming the extension is present).

Probably tls_collect_extensions() and tls_parse_all_extensions() are going to need a new argument to say whether the init/final functions should be called or not.

If we make those changes I think probably your changes in this PR would no longer be required - so I suggest ditching this PR altogether and starting again but with this new approach.